llvm/test/CodeGen/RISCV/rvv/vl-opt-instrs.ll

```diff
@@ -1122,6 +1122,132 @@ define <vscale x 4 x i32> @vrem_vx(<vscale x 4 x i32> %a, i32 %b, iXLen %vl) {
   ret <vscale x 4 x i32> %2
 }
 
+define <vscale x 4 x i64> @vwmul_vv(<vscale x 4 x i16> %a, <vscale x 4 x i16> %b, iXLen %vl) {
+; NOVLOPT-LABEL: vwmul_vv:
+; NOVLOPT:       # %bb.0:
+; NOVLOPT-NEXT:    vsetvli a1, zero, e16, m1, ta, ma
+; NOVLOPT-NEXT:    vwmul.vv v12, v8, v9
+; NOVLOPT-NEXT:    vsetvli zero, a0, e32, m2, ta, ma
+; NOVLOPT-NEXT:    vwmul.vv v8, v12, v12
+; NOVLOPT-NEXT:    ret
+;
+; VLOPT-LABEL: vwmul_vv:
+; VLOPT:       # %bb.0:
+; VLOPT-NEXT:    vsetvli zero, a0, e16, m1, ta, ma
+; VLOPT-NEXT:    vwmul.vv v12, v8, v9
+; VLOPT-NEXT:    vsetvli zero, zero, e32, m2, ta, ma
+; VLOPT-NEXT:    vwmul.vv v8, v12, v12
+; VLOPT-NEXT:    ret
+  %1 = call <vscale x 4 x i32> @llvm.riscv.vwmul.nxv4i64.nxv4i32.nxv4i32(<vscale x 4 x i32> poison, <vscale x 4 x i16> %a, <vscale x 4 x i16> %b, iXLen -1)
+  %2 = call <vscale x 4 x i64> @llvm.riscv.vwmul.nxv4i64.nxv4i32.nxv4i32(<vscale x 4 x i64> poison, <vscale x 4 x i32> %1, <vscale x 4 x i32> %1, iXLen %vl)
+  ret <vscale x 4 x i64> %2
+}
+
+define <vscale x 4 x i64> @vwmul_vx(<vscale x 4 x i16> %a, i16 %b, i32 %c, iXLen %vl) {
+; NOVLOPT-LABEL: vwmul_vx:
+; NOVLOPT:       # %bb.0:
+; NOVLOPT-NEXT:    vsetvli a3, zero, e16, m1, ta, ma
+; NOVLOPT-NEXT:    vwmul.vx v12, v8, a0
+; NOVLOPT-NEXT:    vsetvli zero, a2, e32, m2, ta, ma
+; NOVLOPT-NEXT:    vwmul.vx v8, v12, a1
+; NOVLOPT-NEXT:    ret
+;
+; VLOPT-LABEL: vwmul_vx:
+; VLOPT:       # %bb.0:
+; VLOPT-NEXT:    vsetvli zero, a2, e16, m1, ta, ma
+; VLOPT-NEXT:    vwmul.vx v12, v8, a0
+; VLOPT-NEXT:    vsetvli zero, zero, e32, m2, ta, ma
+; VLOPT-NEXT:    vwmul.vx v8, v12, a1
+; VLOPT-NEXT:    ret
+  %1 = call <vscale x 4 x i32> @llvm.riscv.vwmul.nxv4i32.nxv4i16.i16(<vscale x 4 x i32> poison, <vscale x 4 x i16> %a, i16 %b, iXLen -1)
+  %2 = call <vscale x 4 x i64> @llvm.riscv.vwmul.nxv4i64.nxv4i64.i32(<vscale x 4 x i64> poison, <vscale x 4 x i32> %1, i32 %c, iXLen %vl)
```

Comment:
This is maybe a nit, but would it be easier if we added separate tests for getOperandInfo in vl-opt-op-info.ll, where we have a simple vadd that's being reduced with one user that's the instruction we want to test?

It would also be good to test the path where the EEWs/EMULs don't match, something like


```llvm
define <vscale x 4 x i64> @vwmul_vv_match(<vscale x 4 x i32> %a, <vscale x 4 x i32> %b, <vscale x 4 x i32> %c, iXLen %vl) {
  %x = call <vscale x 4 x i32> @llvm.riscv.vadd(<vscale x 4 x i32> poison, <vscale x 4 x i32> %a, <vscale x 4 x i32> %b, iXLen -1)
  %y = call <vscale x 4 x i64> @llvm.riscv.vwmul(<vscale x 4 x i64> poison, <vscale x 4 x i32> %x, <vscale x 4 x i32> %c,  iXLen %vl)
  ret <vscale x 4 x i64> %y
}

define <vscale x 4 x i64> @vwmul_vv_no_match(<vscale x 4 x i32> %a, <vscale x 4 x i32> %b, <vscale x 8 x i16> %c, iXLen %vl) {
  %x = call <vscale x 4 x i32> @llvm.riscv.vadd(<vscale x 4 x i32> poison, <vscale x 4 x i32> %a, <vscale x 4 x i32> %b, iXLen -1)
  %x.bitcast = bitcast <vscale x 4 x i32> %x to <vscale x 8 x i16>
  %y = call <vscale x 8 x i32> @llvm.riscv.vwmul(<vscale x 8 x i32> poison, <vscale x 8 x i16> %x.bitcast, <vscale x 8 x i16> %c,  iXLen %vl)
  ret <vscale x 4 x i64> %y
}
```

Reply:
I want to make sure I understand the suggestion. Are you saying that vl-opt-instrs.ll should be of the form:

```llvm
define <vscale x 4 x i32> @instruction_i_am_testing(<vscale x 4 x i32> %a, iXLen %vl) {
  %1 = call <vscale x 4 x i32> @llvm.riscv.instruction_i_am_testing(<vscale x 4 x i32> poison, <vscale x 4 x i32> %a, i32 5, iXLen -1)
  %2 = call <vscale x 4 x i32> @llvm.riscv.vadd(<vscale x 4 x i32> poison, <vscale x 4 x i32> %1, <vscale x 4 x i32> %a, iXLen %vl)
  ret <vscale x 4 x i32> %2
}
```

and  vl-opt-op-info.ll should be the opposite:

```llvm
define <vscale x 4 x i32> @instruction_i_am_testing(<vscale x 4 x i32> %a, iXLen %vl) {
  %1 = call <vscale x 4 x i32> @llvm.riscv.vadd(<vscale x 4 x i32> poison, <vscale x 4 x i32> %1, <vscale x 4 x i32> %a, iXLen -1)
  %2 = call <vscale x 4 x i32> @llvm.riscv.instruction_i_am_testing(<vscale x 4 x i32> poison, <vscale x 4 x i32> %a, i32 5, iXLen %vl)
  ret <vscale x 4 x i32> %2
```

plus add tests the path where the EEWs/EMULs don't match?

Reply:
If its okay with you, lets make sure that we have the tests we'd like for integer widening multiply instructions in vl-opt-instrs.ll in this patch, and then in one PR I will split the file for all in vl-opt-instrs.ll?

Reply:
Yes, that sounds good